gateway/integration_tests/src/main.rs

```diff
@@ -544,6 +547,82 @@ async fn config_test(gw_type: LightningNodeType) -> anyhow::Result<()> {
     .await
 }
 
+/// Test that verifies the various liquidity tools (onchain, lightning, ecash)
+/// work correctly.
+async fn liquidity_test() -> anyhow::Result<()> {
+    devimint::run_devfed_test(|dev_fed, _process_mgr| async move {
+        let federation = dev_fed.fed().await?;
+        let gatewayd_version = util::Gatewayd::version_or_default().await;
+        if gatewayd_version < *VERSION_0_5_0_ALPHA {
+            info!(%gatewayd_version, "Version did not support gateway liquidity management, skipping");
+            return Ok(());
+        }
+
+        let gw_lnd = dev_fed.gw_lnd_registered().await?;
+        let gw_cln = dev_fed.gw_cln_registered().await?;
+        let gw_ldk = dev_fed
+            .gw_ldk_connected()
+            .await?
+            .as_ref()
+            .expect("Gateways of version 0.5.0 or higher support LDK");
+        let gateways = [gw_lnd, gw_cln, gw_ldk];
+        let gateway_matrix = gateways
+            .iter()
+            .cartesian_product(gateways.into_iter())
+            .filter(|(a, b)| a.ln_type() != b.ln_type());
+
+        info!("Testing payments between gateways...");
+
+        for (gw_send, gw_receive) in gateway_matrix.clone() {
+            info!(
+                "Testing lightning payment: {} -> {}",
+                gw_send.ln_type(),
+                gw_receive.ln_type()
+            );
+
+            let invoice = gw_receive.create_invoice(1_000_000).await?;
+            gw_send.pay_invoice(invoice).await?;
+        }
+
+        info!("Pegging-in gateways...");
+
+        federation
+            .pegin_gateways(1_000_000, Vec::from(gateways))
+            .await?;
+
+        info!("Testing ecash payments between gateways...");
+        for (gw_send, gw_receive) in gateway_matrix.clone() {
+            info!(
+                "Testing ecash payment: {} -> {}",
+                gw_send.ln_type(),
+                gw_receive.ln_type()
+            );
+
+            let ecash = gw_send.send_ecash(federation.calculate_federation_id(), 500_000).await?;
+            gw_receive.receive_ecash(ecash).await?;
+        }
+
+        info!("Pegging-out gateways...");
+        federation.pegout_gateways(500_000_000, Vec::from(gateways)).await?;
+
+        info!("Testing closing all channels...");
+        for gw in gateways {
+            gw.close_all_channels(dev_fed.bitcoind().await?.clone()).await?;
+            let curr_lightning_balance = gw.get_balances().await?.lightning_balance_msats;
```

Comment:
Should we also assert `GetBalancesResponse.inbound_lightning_liquidity_msats`?

Reply:
Done

Reply:
Nit: We don't need to call `gw.get_balances()` twice here

Reply:
Fixed